### ISSUE TITLE
Dual contouring

### ISSUE THREAD

**Issue by LeLocTai** [fogleman/sdf#7]
Dual contouring is a meshing technique which is better at producing sharp corner than marching cube.

Do you think the technique would be suitable in this case? I've been looking for excuse to play with it. Is it something you already working on? If not I'm calling dibs

**Comment by fogleman** [comment]
Go for it.

**Comment by smilberger** [comment]
@LeLocTai I say go for it too, though you have some other options other than just implementing DC yourself.  

1) If you want to learn the DC algorithms, it *would* be an interesting project to modify this project and implement an alternate DC meshing style using numpy.  It would also be nice to have this code available as Blender plug-in.  I'm thinking about doing an implementation in C to learn--my main problem is finding a good math/linear algebra library like Eigen.

2) If you want to do DC with Python with the least amount of work, there is a project on Github called Libfive which is one of the best DC implementations around.  It has a C API for use with multiple languages and the beginnings of a Python binding.  It also has a preliminary Unity binding.  I've used a Python binding to load meshes into Blender in the console, but it is prototype code.  I wrote a prototype Python viewer in OpenGL.

3) If you want to experiment with a working DC implementation right now, but don't care about the language, the Libfive project has a "live coding" GUI in Guile Scheme.  The author of Libfive has been talking about moving the main scripting language to Python, but the advantage of Guile is that it has a nice sandbox for security.

**Comment by LeLocTai** [comment]
Thank for the suggestions! 

I'm currently trying to integrate this project (Python) https://github.com/BorisTheBrave/mc-dc . Mostly because it come with an excellent article.

Performance is a big issue with this approach, because the algorithm want the normal (equate to at least 3 sdf eval) at arbitrary point in space, not a regular grid. I'm not very familiar with numpy, but I'm not sure if it can accelerate this. I'll look at libfive if I can't make this work.

**Comment by smilberger** [comment]
Thanks for that link, I hadn't seen that one.

Libfive uses octrees instead of an uncompressed grid, flattening the shape tree to a tape, as well as other "compiler-like" optimizations of the shape tree to speed meshing up.  It uses the more advanced manifold DC or iso-simplex techniques.   

Matt Keeter, the project's creator, also has a nice [website](https://mattkeeter.com) with write-ups on the optimizations, QEF implementation, etc.  You could ask Matt as an issue on the Libfive Github project if it would be feasible to do the most time-consuming part in Numpy and the rest in Python--I think he's worked a lot with Python--I haven't.  I would be interested in knowing.

In any case, good luck!

**Comment by fogleman** [comment]
Feel free to re-open / PR if you get something going here.

**Comment by LeLocTai** [comment]
Sorry, something come up and I totally forgot about this. I did got something running, but the quality is not what I expected. Certain features fit better, some worse: (right is DC)
![21-02-22_11 59 59_1272x830](https://user-images.githubusercontent.com/12182348/109744830-87cd6e00-7c05-11eb-84de-440ee29709b6.png)
![image](https://user-images.githubusercontent.com/12182348/109745222-2063ee00-7c06-11eb-924f-9c9a520c760b.png)

I'm not sure whether there's a problem with my implementation or the that how the algorithm is. Also it's super slow atm so I haven't sent a PR. Will come back when I got the time. If anyone else interested, my current progress (ugly) is on my fork
